src/quilla/steps/validations/xpath.py

```diff
@@ -204,7 +210,50 @@ def _check_not_attribute_has_value(self) -> ValidationReport:
     def _check_visual_parity(self) -> ValidationReport:
         self._verify_parameters('baselineID')
 
-        return self._create_report(
-            False,
-            msg='VisualParity Validation State not yet implemented'
+        baseline_id = self.parameters['baselineID']
+
+        treatment_image_bytes = self.element.screenshot_as_png
+        treatment_image = Image.open(BytesIO(treatment_image_bytes))
+
+        plugin_result: List = self.ctx.pm.hook.quilla_get_visualparity_baseline(
+            ctx=self.ctx,
+            baseline_id=baseline_id
+        )
+
+        if len(plugin_result) == 0:
+            return VisualParityReport(
+                success=False,
+                target=self._target,
+                browser_name=self.driver.name,
+                baseline_id=baseline_id,
+                msg='No baseline storage mechanism configured, or no baseline found'
+            )
+
+        baseline_image_bytes, baseline_uri = plugin_result[0]
+        baseline_image = Image.open(BytesIO(baseline_image_bytes))
+
+        success = baseline_image == treatment_image  # Run the comparison with Pillow
+
+        if success:
+            return VisualParityReport(
+                success=success,
+                target=self._target,
+                browser_name=self.driver.name,
+                baseline_id=baseline_id,
+            )
+
+        treatment_uri = self.ctx.pm.hook.quilla_store_image(
+            ctx=self.ctx,
+            baseline_id=baseline_id,
+            image_bytes=treatment_image_bytes,
+            image_type=VisualParityImageType.TREATMENT,
+        )
+
+        return VisualParityReport(
+            success=False,
```

Comment:
The first time VP runs there is no baseline image, so the test will fail. Here we need to pass in treatment image bytes for plugin to generate baseline image. You may also want to notify test owners that the test fails because missing baseline image.

Reply:
I'll double check but I think I have an issue already to address this, since I had a couple of ideas to run past you and Daniel for this, cuz if it's not carefully done it would override the baseline every time